d3log/src/tcp_network.rs

```diff
@@ -0,0 +1,114 @@
+
+    // prefer big endian for generic reasons, but we are travelling through
+    // a constricted u64 pipe
+    u128::from_le_bytes([i[0], i[1], i[2], i[3], p[0], p[1], 0, 0, 0, 0, 0,
+                         0,0,0,0,0])
+}
+
+pub fn nid_to_socketaddr(n:Node) -> SocketAddr {
+    let b = n.to_be_bytes();
+    SocketAddr::new(IpAddr::V4(Ipv4Addr::new(b[0], b[1], b[2], b[3])), u16::from_be_bytes([b[4], b[5]]))
+}
+
+pub struct TcpNetwork {
+    peers : Arc::<Mutex::<HashMap<Node, Arc::<Mutex::<TcpStream>>>>>
+}
+
+
+// we wanted to make a network class, that would have a send function, and
+// that would allow us to encapsulate quite a bit about our environment.
+//
+// there is an async_trait macro that promises to do this, but unsurprisingly
+// MutexGuard does not implement Send. So its unclear how to let such
+// a trait function mutate. punting.
+
+impl TcpNetwork {
+    pub async fn soft_intern(&self, nid:Node) -> Result<Arc::<Mutex::<TcpStream>>, std::io::Error> {
+        let p = &mut (*self.peers).lock().await;
+        
+        match p.get(&nid) {
+            Some(x) => Ok((*x).clone()), 
+            None => {
+                let c = Arc::new(Mutex::new(TcpStream::connect(nid_to_socketaddr(nid)).await?));
+                p.insert(nid, c.clone());
+                Ok(c)
+            }
+        }
+    }
+        
+    pub fn new () -> TcpNetwork {
+        TcpNetwork{peers:Arc::new(Mutex::new(HashMap::new()))}
+    }
+
+    pub fn send(&self, nid:Node, _b:Arc::<Mutex::<Batch>>) ->
+        Result<Pin<Box<dyn Future<Output=Result<(),std::io::Error>> + Send + Sync + '_>>,
+               std::io::Error> {
+            println!("send {}", nid);
+            Ok(Box::pin(async move {
+                let p = &mut (*self.peers).lock().await;
+                let encoded = Vec::<u8>::new();
+                // actually this creates a connection, so...kinda yeah
+                let c = self.soft_intern(nid).await.expect("cant fail");
+                let mut cl = c.lock().await;
+                cl.write_all(&encoded).await
+            }))
+        }
+}
+
+use mm_ddlog::typedefs::d3::Workers;
+
+pub async fn bind(n:&TcpNetwork, _t:Arc<Mutex<TransactionManager>>) -> Result<(), std::io::Error> {
+    let listener = TcpListener::bind("127.0.0.1:0").await?;
+    
+    // there is likely a more direct translation not through strings - reevalute nid
+    let a = listener.local_addr().unwrap();
+    let nid = address_to_nid(a);
+    // should be a better contract about the name(s) of the membership relations, or more broadly
+    // system defined relations
+
+    let w = Workers{x:nid}.into_ddvalue();
+    output_json(&(singleton("d3::Workers", &w)?)).await?;
+    loop {
+        let (socket, a) = listener.accept().await?;
+        output_json(&(singleton("Connection", &(address_to_nid(a) as u64).ddvalue())?)).await?;
+        // xxx - socket errors shouldn't stop the listener
+        process_socket(n, socket).await?;
+    }    
```

Comment:
No exit condition?

Reply:
when would you like to stop listening to new connections? are you think we should have an orderly shutdown button?

Reply:
I would assume we'd want some sort of structured shutdown to do cleanup and all that fun stuff

Reply:
The alternative view based on recovery oriented computing is that you will crash so there is no reason to have a clean showdown path. Just make sure you have a working recovery scheme.

Reply:
I'm kinda with mbudiu. to me the cleanup in this case is primarily just to show that we _can_ do the cleanup (that we did all the accounting)....I can't think of a reason to actually free a structure or thread  which was explicitly allocated to live forever. you can imagine my frustration at waiting 60 seconds for a linux box to 'shut down'.  

but its not a big deal, willing to go w/ consensus here